### ISSUE TITLE
Help with disorderly escape?

### ISSUE THREAD

**Issue by a1diablo** [xphoniex/Google-Foobar#5]
I'm currently stuck on the final level and am pulling my hair out over this one. I've tried to make a brute force and search for patterns in the data but in vain. Also, I've been thinking in terms of a 3 dimensional dp maybe but unable to form an intuition. Can you tell me if I'm going in the right direction, or a hint maybe?

**Comment by xphoniex** [comment]
Relax dude, this one is not easy. This is a **math** problem so forget about trying to solve it the same way you would solve a CS problem.

Here's a tip, draw a 2x2 board, write down all the ways pieces can be transformed and where the original cells end up in the transformation, that should give you some information about how many combinations there is.

**Comment by xphoniex** [comment]
^try spending a few days on above. You're very smart if you can solve it. If you're stuck still I'll give you another tip ;)

**Comment by varadiz** [comment]
Stuck as well. 
My logic for w=2, h=2, s=2:

STEP1. Find all possible combinations of "how many times each number occurs"
{0: 4, 1: 0} -> there are 4 zeros and 0 ones.
{0: 1, 1: 3} -> there is 1 zero and 3 ones.
{0: 3, 1: 1} -> there are 3 zeros and 1 one.
{0: 2, 1: 2} -> there are 2 zeros and  2 ones.

STEP2. For each of  these, check possible column combinations.
  Example1:
  {0: 0, 1: 4}
  Possible columns are:  (1, 1), (1, 1)
  meaning 1st column contains 2 ones, second column also contains 2 ones.

Example2:
  {0: 2, 1: 2}
  Possible columns are:
  (0, 1), (0, 1) -> meaning 1st column has 1 zero and 1 one, same in the 2nd column.
  (0, 0), (1, 1) -> meaning 1st column has 2 zeros and 2nd column has 2 ones.

STEP3. For each of these column possibilities, I used itertools.product to get possible row combinations.
Eg.:
 (0,1), (0,1) -> will give 00, 01 and 11 as possible rows. 

STEP4. Next, I looked at all possible combinations for row possibilities to give the desired w*h numbers using itertools.combinations_with_replacement. I also checked if the resulting "matrix" has the same number of ones and zeros as desired and if the resulting columns would match occurrence of each number. 

So to sum up, {0: 2, 1: 2} (2 zeros, 2 ones) will give two possible column combination:
(0,1), (0,1) -> 1st column and 2nd column both have 1 zero and 1 one.
  Possible matrix:
  00   
  11              

  01
  10

(0,0), (1,1) -> 1st column has 2 zeros and 2nd column has 2 ones.
  Possible matrix:
  01
  01

COMMENTS
This method is brute forcing like crazy and I can't really see a way to make some shortcuts.
It gives the correct answer ("7") for test case 1, but fails for test case 2, returning "434" instead of "430". Still trying to figure out where I double count. 
For bigger arguments, it takes forever. So long, that I can't even verify it because foobar gives me the Time exceeded error. 

Anyway, could you give me some hint if I'm going in the right direction? Did you have a totally different approach? 
Any help would be much appreciated, I dream about this stuff now, it's driving me crazy.

**Comment by a1diablo** [comment]
As @dave said, this is not to be solved through computer science methods. When it falls short answers often fall in maths. If you're not a major in maths, as I'm not, you'll need to study some stuff before getting your hands on this. Things I would suggest is: group theory, Burnside lemma, and polya enumeration theorem. Go ahead and absorb these concepts. It's totally worth it. I learnt a lot of new things in the process and on my way to solve this one. Good luck (y)

**Comment by varadiz** [comment]
Thank you! I'll go over them.

**Comment by devojoyti** [comment]
Hi @abhimanyu-bst , I went through Burnside lemma and Polya Enumeration, and in my opinion, the polynomial I am getting is this - 

(  s^(wh) +  wC2 * s^(h(w - 1)) + hC2 * s^(w(h - 1)) + (wC2)*(hC2) * ( ??? ) ) / (w! * h!) 

Where, 

(w! * h!) = |G|, as it is different number of configurations possible by different combinations of row and column swap. 

(s ^ (w*h)) -> Base case, all possible combinations

wC2 * s^(h*(w - 1)) -> Number of similar config. after any two row swap

hC2 * s^(w*(h - 1)) -> Number of similar config. after any two column swap

(wC2)*(hC2) * ( ??? ) ) -> ? Supposed to be number of similar config. after row and column swaps. This one I am not getting.....

Can you please help me ? Been struck on this by 8 days now :(

**Comment by foobar112358** [comment]
Hello Dave,

I was pretty sure that I found the correct formula for the number of orbits using Burnside's Lemma, for which I count the number of permutations in S_w * S_h with a given number of disjoint cycles, say c, as each of these will fix a set of cardinality s^c.  I enumerated the permutations for each c via integer partitions and basic number theory.  

I will give you the details of my approach so that you can see where I'm stuck, but I would prefer to send them to you in a private message if possible.  Failing that, I will add them here, but I don't want to spoil the problem for anyone else.  

I must have neglected something important in my calculations or implementation, as I consistently fail test cases 4 through 10.  As I **only have a week left** to finish the problem, I am at the point where I could really use a timely hint about what I am missing.  If you are willing to provide said hint, please let me know the best way to give you the relevant details of what I've done.  Thank you!

**Comment by xphoniex** [comment]
@foobar112358 what is your output for s = 20, h = 12, w = 12?

**Comment by foobar112358** [comment]
I have that answer(12, 12, 20) is:

97195340925396730736950973830781340249131639866983741676036180016052320403466942623734036310122017767061431378848635938035448728104820484623594701991751592717158362452044

**Comment by xphoniex** [comment]
@foobar112358 it should be:

971953409253967307369509738307813402491316**79073592360856141700148734207997877978005419735822878768821088343977969209139721682171487959967012286474628978470487193051591840**

You might be losing precision somewhere. This should help you determine where:

22300745198531341817894657615847583802746155678399565967993124561615436149415899871380763938002071429362749343044868906014253019900488985895312634973607213920301808604240055624427110400000 **divided by** 229442532802560000

**Comment by foobar112358** [comment]
Thanks so much, @xphoniex !  I really appreciate your hint.

Alas, however, I still can't figure out where I could be losing precision.  I'm using Python, which, if I understand correctly, should handle very large integers exactly for the operations +, -, *, and //.  But I'm new to Python, so I might be making assumptions that aren't true.  I did go back and replace my s**c with a naive power function power(s, c) that I wrote into the code, but that didn't make any difference in my output, so my problem seems not to be exponentiation per se (although the underlying multiplication might be off if I've misunderstood how to use large integers in Python).

So, I'm trying to figure out whether I am indeed losing precision, or whether I have a logic error in my implementation, or even whether the enumeration formula that I derived is not quite right.  That being said, I am fairly certain of my formula, having compared its elements to information found online (and having specialized in combinatorics for my degree!)...but of course it still could be wrong.

Anyway, long story short, I'm still stuck.  Would it be possible for you to tell me whether or not the following output is already incorrect?

answer(3, 5, 20):

45568090499534008

I wrote a C++ program that was purely brute force.  My Python output agrees with the brute-force output, but I couldn't get any higher than answer(2, 3, 4) using plain brute force.  So if I do have some kind of logic error, it doesn't seem to affect my output until after that point, which is why I'm mystified. 

Anyway, thanks again!

**Comment by xphoniex** [comment]
@foobar112358 For 3x5 with 20 states your output is correct. 

I'll give you a few more samples for debugging purpose:

```
(4,4,20) -> 1137863754106723400
(5,5,20) -> 23301834615661488487765745000
(6,6,20) -> 132560781153101038829213988789736592649360
(7,7,20) -> 221619886894198821201872678876163305792210161226545392840
(8,8,20) -> 113469378614817897312718329989374518983724697432844009920312263602471667640
```

Most likely you won't get the same output from 7x7. You can easily lose precision if you're not careful, rule that out before moving into logical debugging.

**Comment by foobar112358** [comment]
Thank you SO much!!!

With the samples you provided, I was able to narrow down where my program stopped producing correct output.  There was a logic error in the function I had written to produce all of the partitions of an integer.  The error made my function skip over some valid partitions for 6 onwards.  One little tweak---literally cutting and pasting a single condition from one line to another---and voilà!

Thanks again!  I really appreciate it!

**Comment by hirenvasani** [comment]
@foobar112358 @xphoniex Even I am badly stuck with in between. Can you guys share some more hint. I am able to get devided by 229442532802560000 value in case of answer(12, 12, 20).

**Comment by xprilion** [comment]
@foobar112358 @xphoniex some more hints? :( just a week left for me and I can't get the head and tails of this question...I'd not studied group theory before this...and I am really bad at applying it so far :p

**Comment by hirenvasani** [comment]
@xphoniex @foobar112358 @abhimanyu-bst @Data-i-Consulting @devojoyti I finally got breakthrough can you tell me what is the answer of **(3,3,3) is it 26568/36** i.e. **738**

**Comment by rixeli** [comment]
can you please provide solution as well?

**Comment by xprilion** [comment]
I solved all through the foobar! 😀 

@hirenvasani yes the answer is 738.

@vimalkrpal what would be the fun of solving foobar then?

Some test cases:

2, 2, 2 - 7
3, 3, 3 - 738
3, 3, 4 - 8240
3, 3, 5 - 57675
4, 4, 4 - 7880456
4, 4, 5 - 270656150
5, 5, 5 - 20834113243925

**Comment by Pancham97** [comment]
Hi All,

I have been stuck on this problem too. It's driving me crazy. I've tried to read, watch every material available on Burnside Lemma, Polya Enumeration Theorem and cycle indices. But, I am still unable to get a hang of it. Could you please help? I'd really appreciate you taking some time. Thanks!

**Comment by PCDSandwichMan** [comment]
This method has been registered as a frequent answer and it gets flagged as cheating so don't actually use this but you can use it to verify what the output should be since no one seems to be responding here anymore.
`
from math import factorial, gcd
from collections import Counter

def cycle_counter(cycle, block):
    cycle_count = factorial(block)
    for a, b in Counter(cycle).items():
        cycle_count //= (a**b)*factorial(b)
    return cycle_count

def gird_maker(block, i=1):
    yield [block]
    for i in range(i, block//2 + 1):
        for j in gird_maker(block-i, i):
            yield [i] + j
            
def solution(w, h, s):
    grid_count = 0 
    for cycle_w in gird_maker(w):
        for cycle_h in gird_maker(h):
            cycle_count = cycle_counter(cycle_w, w)*cycle_counter(cycle_h, h)
            grid_count += cycle_count * \
                (s**sum([sum([gcd(i, j) for i in cycle_w]) for j in cycle_h]))

    return grid_count // (factorial(w)*factorial(h))
`

**Comment by hamza-mughees** [comment]
> Hi @abhimanyu-bst , I went through Burnside lemma and Polya Enumeration, and in my opinion, the polynomial I am getting is this -
> 
> ( s^(wh) + wC2 * s^(h(w - 1)) + hC2 * s^(w(h - 1)) + (wC2)*(hC2) * ( ??? ) ) / (w! * h!)
> 
> Where,
> 
> (w! * h!) = |G|, as it is different number of configurations possible by different combinations of row and column swap.
> 
> (s ^ (w*h)) -> Base case, all possible combinations
> 
> wC2 * s^(h*(w - 1)) -> Number of similar config. after any two row swap
> 
> hC2 * s^(w*(h - 1)) -> Number of similar config. after any two column swap
> 
> (wC2)*(hC2) * ( ??? ) ) -> ? Supposed to be number of similar config. after row and column swaps. This one I am not getting.....
> 
> Can you please help me ? Been struck on this by 8 days now :(

Hi, have you been able to figure out the full equation?

**Comment by shivampapat** [comment]
Can anyone help me with java code??
It is running on my machine with all cases I put even the one discussed above in this thread. It is giving accurate output but showing test case failed on verifying on foobar.
